### PR TITLE
[c10d] dump from one and only one thread (PG0's monitor thread)

### PR DESCRIPTION
**Pull request by shuqiangzhang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120893

Summary:
When there are multiple PGs in a process and a hardware failure happens,
we found that multiple PGs/ threads in the same
process are competing to dump the same records at the same time. The
affects the reliability of dumps.

In this PR, we will try to make the change such that only one thread/PG
could dump: PG0's monitor thread. We use a static variable to indicate
that something (e.g., collective timeout) has triggered the dump
locally.

monitor thread would dump debug info under any one of the 3 conditions:
1: this static variable is set to true by the watchdog thread when it detects
a timeout or pipe dump signal
2: timeout signal is received from other ranks through tcpstore
3: no heartbeat of watchdog
Test Plan:
python test/distributed/test_c10d_nccl.py -k
test_timeout_dumps_on_stuck_ranks

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -1553,21 +1580,11 @@ void ProcessGroupNCCL::watchdogHandler() {
...
-                  "Flight recorder dump in watchdog");
-              // This sleep is used to give additional time for other ranks to
-              // receive the dump signal and then dump flight records, the same
-              // sleep time used in the heartbeatmonitor thread
-              // TODO, we should probably dump from a single thread in the
-              // future
+            if (dumpOnTimeout_) {
+              // signal the monitor thread to start dumping
+              shouldDump_.store(true);
+              // This sleep is used to give time for dumping before throwing
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1586]: i think we could also consider never throwing exceptions or commAborting from watchdog thread, and, using a similar means (perhaps storing a static exception ptr if we think of a better way) letting the monitor thread assume sole responsibility for 'exiting the process' if it's signaled to do so

could be a separate PR. wdyt?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -1197,6 +1200,32 @@ void ProcessGroupNCCL::heartbeatMonitor() {
...
+            ".");
+        exitMsg = c10::str(
+            "ProcessGroupNCCL's watchdog detected a collective timeout from the local rank. ",
+            "This is most likely caused by incorrect usages of collectives, e.g., wrong ",
+            "sizes used across ranks, the order of collectives is not same for all ranks ",
+            "or the scheduled collective, for some reason, didn't run. Additionally, ",
+            "this can be caused by GIL deadlock or other reasons such as network errors or ",
+            "bugs in the communications library (e.g. NCCL), etc. We tried our best to ",
+            "dump the debug info into the storage to help you debug the issue.");
+        break;
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1227]: does this break mean 'exit the monitor loop and kill the job?'

bc that isn't safe.  When the 'dumpPipe' is invoked, we expect the dump signal to propagate here and dump, but the training should continue.  Only if the dump is caused by timeout we want to begin the teardown.

Comment (shuqiangzhang) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1227]: Oh, didn't know dumpPipe is for dump and continue

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (shuqiangzhang) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1586]: Yes, probably a single place to exit as well, will try it in another PR

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1227]: yea. its probably worth adding a test that after dumpPipe is written the process doesn't die. i thought we had that actually.

Comment (shuqiangzhang) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1227]: We have the test, but I misunderstood the dump logic, so to disable the Kill part through: os.environ['TORCH_NCCL_ENABLE_MONITORING'] = '0', in the next version, I will remove this env change

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -1406,59 +1443,6 @@ std::string ProcessGroupNCCL::getNCCLWatchdogDebugInfo() {
   return retrieveDesyncReport(store_, "NCCL", rank_, size_);
 }
 
-#if defined(__linux__)
-struct DumpPipe {
```
Comment (shuqiangzhang) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1410]: Move the declaration to the header file so that heartbeatMonitor() function can access it, defining in cpp file will make the access depending on the order of function implementation

**Review by wconstab — APPROVED** [review]
lgtm. thanks!
